### ISSUE TITLE
Pod in ContainerCreating status when containers.resources.limits.memory use lowercase unit 

### ISSUE THREAD

**Issue by monday0rsunday** [kubernetes/kubernetes#49442]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Pod is in eventually *ContainerCreating* status

```
$ kubectl describe pod limiting-redis
Name:		limiting-redis
Namespace:	default
Node:		kubectl/192.168.2.11
Start Time:	Sat, 22 Jul 2017 13:51:29 +0000
Labels:		<none>
Annotations:	<none>
Status:		Pending
IP:
Containers:
  redis-container:
    Container ID:
    Image:		kubernetes/redis:v1
    Image ID:
    Port:		<none>
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Limits:
      memory:	500m
    Requests:
      memory:		500m
    Environment:	<none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-k0pqt (ro)
Conditions:
  Type		Status
  Initialized 	True
  Ready 	False
  PodScheduled 	True
Volumes:
  default-token-k0pqt:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-k0pqt
    Optional:	false
QoS Class:	Burstable
Node-Selectors:	<none>
Tolerations:	node.alpha.kubernetes.io/notReady:NoExecute for 300s
		node.alpha.kubernetes.io/unreachable:NoExecute for 300s
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  57s		57s		1	default-scheduler			Normal		Scheduled	Successfully assigned limiting-redis to kubectl
  56s		3s		42	kubelet, kubectl			Normal		SandboxChanged	Pod sandbox changed, it will be killed and re-created.
  57s		2s		43	kubelet, kubectl			Warning		FailedSync	Error syncing pod, skipping: failed to "CreatePodSandbox" for "limiting-redis_default(dca772d6-6ee4-11e7-838f-02ea070db58b)" with CreatePodSandboxError: "CreatePodSandbox for pod \"limiting-redis_default(dca772d6-6ee4-11e7-838f-02ea070db58b)\" failed: rpc error: code = 2 desc = failed to start sandbox container for pod \"limiting-redis\": Error response from daemon: {\"message\":\"invalid header field value \\\"oci runtime error: container_linux.go:247: starting container process caused \\\\\\\"process_linux.go:286: decoding sync type from init pipe caused \\\\\\\\\\\\\\\"read parent: connection reset by peer\\\\\\\\\\\\\\\"\\\\\\\"\\\\n\\\"\"}"
```

**What you expected to happen**:

There should be invalid response when I create pod
or Pod should be in eventually Running status

**How to reproduce it (as minimally and precisely as possible)**:

Create a pod with containers.resources.limits.memory using lowercase unit(500m, 1000m, etc.), e.g.

```
apiVersion: v1
kind: Pod
metadata:
    name: limiting-redis
spec:
    containers:
    - name: redis-container
      image: kubernetes/redis:v1
      resources:
        limits:
          memory: "500m"
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.2", GitCommit:"922a86cfcd65915a9b2f69f3f193b8907d741d9c", GitTreeState:"clean", BuildDate:"2017-07-21T08:23:22Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.6-rancher1", GitCommit:"973d2fa162b887a6f696f4574b85e949d480a071", GitTreeState:"clean", BuildDate:"2017-06-17T03:54:44Z", GoVersion:"go1.7.6", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration**: R
- OS (e.g. from /etc/os-release):
```
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```
- Kernel (e.g. `uname -a`): `Linux kubectl 4.4.0-77-generic #98-Ubuntu SMP Wed Apr 26 08:34:02 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
- Others:

**Comment by xiangpengzhao** [comment]
/sig node

Do you try to use `500Mi` and then see if the container can start? 

Don't use lowercase memory unit. See: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory

However, IIRC, even though `m` is an invalid memory unit, if you use it, it will be translate to milli byte.

/cc @vishh

**Comment by monday0rsunday** [comment]
@xiangpengzhao yes, container can start with 500Mi

> IIRC, even though m is an invalid memory unit, if you use it, it will be translate to milli byte

I think that there should be clear documentation, or invalid response

**Comment by supereagle** [comment]
For memory, the unit should be `M` or `Mi`, not `m`, `m` is the unit for CPU.

I think there is need to validate the quantity unit, and I will submit a PR for this.

**Comment by supereagle** [comment]
Correct quantities:
* resource.Quantity{i:resource.int64Amount{value:524288000, scale:0}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"", Format:"DecimalSI"}
* resource.Quantity{i:resource.int64Amount{value:500, scale:6}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"500e6", Format:"DecimalExponent"}
* resource.Quantity{i:resource.int64Amount{value:524288000, scale:0}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"500Mi", Format:"BinarySI"}

Error quantity:
* resource.Quantity{i:resource.int64Amount{value:500, scale:-3}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"500m", Format:"DecimalSI"}

Not sure directly forbid the unit `m` for memory, or convert it to a plain integer.

**Comment by vishh** [comment]
I think adding `ResourceMemory` to [this list of integer resources](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/helper/helpers.go#L193) should fix the problem.

**Comment by supereagle** [comment]
`m` is OK to used as memory unit, it will be treated as 1/1000 byte.

The root cause is that minimum memory limit allowed for Docker is 4MB. If the specified memory limit is less than 4MB, the containers can not be created, so the pod is always in `ContainerCreating` status.

You can refer to [Docker Doc](https://docs.docker.com/engine/admin/resource_constraints/#limit-a-containers-access-to-memory), or verify by the command:
```
# docker run -m 128 nginx:latest
/usr/bin/docker-current: Error response from daemon: Minimum memory limit allowed is 4MB.
```
The solution is to validate the memory limit to ensure that it is not less than 4MB. I will summit the PR to fix this.

**Comment by vishh** [comment]
Docker is not the only runtime. What version of docker are you running? Is it still the case that `4MB` is the minimum requirement for docker? I'd be surprised and I will try to fix docker engine rather than make it a requirement in k8s API. I have gone down that path in the past and abandoned it.

**Comment by xiangpengzhao** [comment]
>The solution is to validate the memory limit to ensure that it is not less than 4MB. I will summit the PR to fix this.

Sounds reasonable. But what if docker changes its minimum memory limit? How can validation know that?

**Comment by monday0rsunday** [comment]
> What version of docker are you running

**Comment by monday0rsunday** [comment]
> What version of docker are you running

My Docker version is 1.12.6.

```
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   78d1802
 Built:        Tue Jan 31 23:35:14 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   78d1802
 Built:        Tue Jan 31 23:35:14 2017
 OS/Arch:      linux/amd64
```

**Comment by supereagle** [comment]
My Docker version is also v1.12.6.

I think this limitation is not set by Docker, but the requirement of Linux kernel.
When set the memory limit to `5000000m`, which is less than 4MB, the created container will be immediately stopped because of oom-killer:
```
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e systemd: Started docker container c79105be6f30a30c1a7a2eee69cf3b3adac9f7387ce44d75ece1bc2d8f615b38.
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e systemd: Starting docker container c79105be6f30a30c1a7a2eee69cf3b3adac9f7387ce44d75ece1bc2d8f615b38.
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: exe invoked oom-killer: gfp_mask=0xd0, order=0, oom_score_adj=-999
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: exe cpuset=docker-c79105be6f30a30c1a7a2eee69cf3b3adac9f7387ce44d75ece1bc2d8f615b38.scope mems_allowed=0
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: CPU: 0 PID: 15826 Comm: exe Not tainted 3.10.0-327.el7.x86_64 #1
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: Hardware name: Red Hat Inc. OpenStack Nova, BIOS 0.5.1 01/01/2007
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: ffff88030e0b5080 00000000ac2afb13 ffff880168057cd0 ffffffff816351f1
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: ffff880168057d60 ffffffff81630191 ffff880168057d80 ffffffff810bdf62
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: ffff880168057d30 ffff880168057d30 ffffffff8116c956 ffff88030e0b55e8
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: Call Trace:
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff816351f1>] dump_stack+0x19/0x1b
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff81630191>] dump_header+0x8e/0x214
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff810bdf62>] ? select_task_rq_fair+0x552/0x6f0
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff8116c956>] ? find_lock_task_mm+0x56/0xc0
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff8116cdee>] oom_kill_process+0x24e/0x3b0
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff81088dae>] ? has_capability_noaudit+0x1e/0x30
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff811d3b05>] mem_cgroup_oom_synchronize+0x555/0x580
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff811d2ef0>] ? mem_cgroup_charge_common+0xc0/0xc0
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff8116d664>] pagefault_out_of_memory+0x14/0x90
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff8162e592>] mm_fault_error+0x8e/0x180
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff816410b1>] __do_page_fault+0x3e1/0x420
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff81641113>] do_page_fault+0x23/0x80
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [<ffffffff8163d408>] page_fault+0x28/0x30
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: Task in /kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podd559c1b5_7c0c_11e7_9d4a_fa163ed438e9.slice/docker-c79105be6f30a30c1a7a2eee69cf3b3adac9f7387ce44d75ece1bc2d8f615b38.scope killed as a result of limit of /kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podd559c1b5_7c0c_11e7_9d4a_fa163ed438e9.slice
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: memory: usage 8kB, limit 8kB, failcnt 81
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: memory+swap: usage 8kB, limit 9007199254740991kB, failcnt 0
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: kmem: usage 0kB, limit 9007199254740991kB, failcnt 0
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: Memory cgroup stats for /kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podd559c1b5_7c0c_11e7_9d4a_fa163ed438e9.slice: cache:0KB rss:0KB rss_huge:0KB mapped_file:0KB swap:0KB inactive_anon:0KB active_anon:0KB inactive_file:0KB active_file:0KB unevictable:0KB
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: Memory cgroup stats for /kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podd559c1b5_7c0c_11e7_9d4a_fa163ed438e9.slice/docker-c79105be6f30a30c1a7a2eee69cf3b3adac9f7387ce44d75ece1bc2d8f615b38.scope: cache:0KB rss:8KB rss_huge:0KB mapped_file:0KB swap:0KB inactive_anon:0KB active_anon:4KB inactive_file:0KB active_file:0KB unevictable:0KB
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [ pid ]   uid  tgid total_vm      rss nr_ptes swapents oom_score_adj name
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: [15826]     0 15826    30491      772      20        0          -999 exe
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: Memory cgroup out of memory: Kill process 15830 (exe) score 384500 or sacrifice child
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e kernel: Killed process 15826 (exe) total-vm:121964kB, anon-rss:780kB, file-rss:2308kB
Aug  8 15:40:29 zhouyue24-7ec4162b-b85b-483f-beac-65678a67a76e systemd: Stopped docker container c79105be6f30a30c1a7a2eee69cf3b3adac9f7387ce44d75ece1bc2d8f615b38.
```

When set the memory limit to `5M`(4.77 MiB), which is a little more than 4MB, there is no problem:
```
# cat redis.yaml
apiVersion: v1
kind: Pod
metadata:
    name: limiting-redis
spec:
    containers:
    - name: redis-container
      image: nginx:latest
      resources:
        limits:
          memory: "5000000000m"  # "524288000"
# kubectl --namespace test get pods
NAME             READY     STATUS    RESTARTS   AGE
limiting-redis   1/1       Running   0          3h
# kubectl --namespace test describe pod limiting-redis
Name:		limiting-redis
Namespace:	test
Node:		10.213.33.245/10.213.33.245
Start Time:	Tue, 08 Aug 2017 11:54:59 +0800
Labels:		<none>
Annotations:	<none>
Status:		Running
IP:		172.30.15.2
Containers:
  redis-container:
    Container ID:	docker://6abbb9c46a0489051ae7e8fbad40b748dc2dce00dc89736984b2bdef1fb5c17a
    Image:		nginx:latest
    Image ID:		docker-pullable://docker.io/nginx@sha256:788fa27763db6d69ad3444e8ba72f947df9e7e163bad7c1f5614f8fd27a311c3
    Port:		<none>
    State:		Running
      Started:		Tue, 08 Aug 2017 11:55:03 +0800
    Ready:		True
    Restart Count:	0
    Limits:
      memory:	5M
    Requests:
      memory:		5M
    Environment:	<none>
    Mounts:		<none>
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	True 
  PodScheduled 	True 
Volumes:	<none>
QoS Class:	Burstable
Node-Selectors:	<none>
Tolerations:	<none>
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  3h		43s		173	kubelet, 10.213.33.245			Warning		MissingClusterDNS	kubelet does not have ClusterDNS IP configured and cannot create Pod using "ClusterFirst" policy. Falling back to DNSDefault policy.
# docker ps
CONTAINER ID        IMAGE                                                                                     COMMAND                  CREATED             STATUS              PORTS               NAMES
6abbb9c46a04        docker.io/nginx@sha256:788fa27763db6d69ad3444e8ba72f947df9e7e163bad7c1f5614f8fd27a311c3   "nginx -g 'daemon off"   3 hours ago         Up 3 hours                              k8s_redis-container_limiting-redis_test_59291d9c-7bed-11e7-9d4a-fa163ed438e9_0
a4760d69f71c        registry.access.redhat.com/rhel7/pod-infrastructure:latest                                "/pod"                   3 hours ago         Up 3 hours                              k8s_POD_limiting-redis_test_59291d9c-7bed-11e7-9d4a-fa163ed438e9_0
# docker stats 6abbb9c46a04
CONTAINER           CPU %               MEM USAGE / LIMIT      MEM %               NET I/O             BLOCK I/O           PIDS
6abbb9c46a04        0.00%               3.328 MiB / 4.77 MiB   69.78%              1.156 kB / 648 B    74.89 MB / 0 B      0
```

**Comment by supereagle** [comment]
> Sounds reasonable. But what if docker changes its minimum memory limit? How can validation know that?

As this limitation is not from Docker, but from Linux kernel, I will read Docker source code to find the source of the minimum memory limit.

**Comment by supereagle** [comment]
The minimum memory limit is a const in [daemon/daemon_unix.go](https://github.com/moby/moby/blob/master/daemon/daemon_unix.go#L56-L57): 
```
	// It's not kernel limit, we want this 4M limit to supply a reasonable functional container
	linuxMinMemory = 4194304
```
Related issue: moby/moby#18405

There is no a good way to know that Docker changes its minimum memory limit.

**Comment by zhangxiaoyu-zidif** [comment]
I have a problem, I set memory to 0.1Gi, that's converted to memory: 107374182400m.
Amazing followed: I cannot set replicas for this rc. That means: I want to scale it to 2 replicas, there is still one running pod. During this processing, several pods were created, but all were deleted soon by dockerd.

have you seen it before?
cc @xiangpengzhao @vishh @supereagle

**Comment by kamesh2229** [comment]
Hi,

I tried deploying using oc debug -f <template name> and getting the below error.

oc debug -f elasticsearch-template.yml
error: cannot debug elasticsearch: unable to extract pod template from type *template.Template

So deployed the normal way how i was doing and  still facing the problem but now the container is not at all started in this case 

oc process -f elasticsearch-template.yml  PROJECT_NAME=sampleName APPLICATION_NAME=elasticsearch | oc apply -f -
  

Events:
  Type     Reason                    Age                From                                     Message
  ----     ------                    ----               ----                                     -------
  Normal   Scheduled                 5m                 default-scheduler                        Successfully assigned elasticsearch-master-7865cd56cd-wdhgx to server
  Normal   SuccessfulMountVolume     5m                 kubelet,                                 MountVolume.SetUp succeeded for volume "storage"
  Normal   SuccessfulMountVolume     5m                 kubelet,                                 MountVolume.SetUp succeeded for volume "default-token-tww8k"
  Warning  FailedCreatePodContainer  10s (x25 over 5m)  kubelet,                                 unable to ensure pod container exists: failed to create container for /kubepods/burstable/pode99ea512-8357-11e8-8880-005056a07a77 : dbus: connection closed by user


oc get pods
NAME                                    READY     STATUS     RESTARTS   AGE
elasticsearch-client-7754b98dc6-z6q89   0/1       Init:0/3   0          29m
elasticsearch-data-77c8b59978-8trjh     0/1       Init:0/3   0          29m
elasticsearch-master-7865cd56cd-6d489   0/1       Init:0/3   0          29m
elasticsearch-master-7865cd56cd-wdhgx   0/1       Init:0/3   0          29m


Tried to get some debug logs by attaching the pod but it is hung in waiting to pod start status.

oc debug elasticsearch-client-7754b98dc6-z6q89
Defaulting container name to elasticsearch-client.
Use 'oc describe pod/elasticsearch-client-7754b98dc6-z6q89-debug -n default-project' to see all of the containers in this pod.
Debugging with pod/elasticsearch-client-7754b98dc6-z6q89-debug, original command: <image entrypoint>
Waiting for pod to start ...

Not sure if something wrong is done here ...
